### PR TITLE
Add `runOnAll` and `runOnNode(address)` to SQLiteClusterMessenger

### PR DESCRIPTION
**Pull request by rafecolton**
### Details

Adds the ability to broadcast commands and run commands on specific nodes using the private command port instead of the node port. This in turn will allow these commands to be handled like any other instead of using the special node port logic. The overall goal with this change is to make all commands operate the same way so they are easy to reason about and parallelize. These new functions in particular will be used extensively in Logrock.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/208103

### Tests

Updated `BroadcastCommand` to check that the commands get broadcast to all of the nodes and not just the first one. Otherwise, just existing tests. In particular, I used:
* `BroadcastCommand`
* `Escalate`
* `GracefulFailover` - this uses the log lines in [8fde3d23cd5809b0381c682be97b5533eb813afc](https://github.com/Expensify/Bedrock/pull/1303/commits/8fde3d23cd5809b0381c682be97b5533eb813afc) to make sure code is being reach in the socket pool as expected. You can revert the revert if you want to try that yourself.

_________
**Internal Testing Reminder:** when changing bedrock, please compile auth against your new changes

### CODE REVIEW

File: BedrockServer.cpp
Comment (rafecolton) [inline BedrockServer.cpp:1927]: Note: This code is just copy/pasted from what used to be in `acceptCommands()`

File: BedrockServer.cpp
Comment (rafecolton) [inline BedrockServer.cpp:2066]: Honest to god could not think of a better variable name today, open to suggestions. We already use `command` and `cmd` and this is all type juggling

**Comment by rafecolton** [comment]
Unsure if more tests are needed, so please comment on that. Also please be nitpicky with style, I want to learn.

File: sqlitecluster/SQLiteClusterMessenger.cpp
Comment (tylerkaraszewski) [inline sqlitecluster/SQLiteClusterMessenger.cpp:255]: Comment no longer fits.

File: sqlitecluster/SQLiteClusterMessenger.h
Comment (tylerkaraszewski) [inline sqlitecluster/SQLiteClusterMessenger.h:32]: Document what is returned here (and for the other new functions below)

File: sqlitecluster/SQLiteClusterMessenger.h
Comment (tylerkaraszewski) [inline sqlitecluster/SQLiteClusterMessenger.h:74]: This doesn't really need to be wrapped in a `unique_ptr` if it's always onwed by the same object and has a very specific lifetime (same as the owning object). you could just use `SMultiSocketPool _socketPool` and it'd be simpler.

File: sqlitecluster/SQLiteNode.cpp
Comment (tylerkaraszewski) [inline sqlitecluster/SQLiteNode.cpp:1972]: I think the `TODO` is to remove this, right? I think the intention is to not use `SQLiteNode::broadcast` any longer.

**Review by tylerkaraszewski — CHANGES_REQUESTED** [review]
First pass complete. It actually looks pretty good. Comments are mostly about making it cleaner or clearer, not general correctness.

File: BedrockServer.cpp
Comment (rafecolton) [reply BedrockServer.cpp:2066]: Thank you, I was blanking so hard when I wrote this, this is great 😆

File: libstuff/SMultiSocketPool.cpp
Comment (rafecolton) [reply libstuff/SMultiSocketPool.cpp:10]: I think what it does is also pretty self-evident, so I'll just delete the comment

File: libstuff/SMultiSocketPool.h
Comment (rafecolton) [reply libstuff/SMultiSocketPool.h:5]: Ah yeah fair enough

File: sqlitecluster/SQLiteClusterMessenger.h
Comment (rafecolton) [reply sqlitecluster/SQLiteClusterMessenger.h:61]: Yeah `static` seems good 👍🏼

File: sqlitecluster/SQLiteNode.cpp
Comment (rafecolton) [reply sqlitecluster/SQLiteNode.cpp:1972]: Oh you know what, I think we can actually remove the function now. We already refactored BedrockServer not to use this in broadcasting, and nothing else calls this directly, anything external like Auth uses `server->broadcastCommand()`

File: sqlitecluster/SQLiteClusterMessenger.h
Comment (rafecolton) [reply sqlitecluster/SQLiteClusterMessenger.h:61]: Oh hm, not sure if this will work, it seems to be complaining about calling member functions and using member variables in the function. I think `const` will work, but not 100% sure yet, got some other complier errors.

File: sqlitecluster/SQLiteClusterMessenger.cpp
Comment (rafecolton) [reply sqlitecluster/SQLiteClusterMessenger.cpp:90]: When I do this and try to compile, I get the following error:

```
sqlitecluster/SQLiteClusterMessenger.cpp:91:34: error: passing ‘const std::map<std::__cxx11::basic_string<char>, SString, STableComp>’ as ‘this’ argument discards qualifiers [-fpermissive]
```

It refers to the line below that does `data["name"]` - any idea how to fix this?

File: sqlitecluster/SQLiteClusterMessenger.h
Comment (rafecolton) [reply sqlitecluster/SQLiteClusterMessenger.h:61]: When I try `const`, I get this error and a handful of other ones:

```
sqlitecluster/SQLiteClusterMessenger.cpp: In member function ‘bool SQLiteClusterMessenger::_sendCommandOnSocket(STCPManager::Socket&, BedrockCommand&) const’:
sqlitecluster/SQLiteClusterMessenger.cpp:148:75: error: passing ‘const SQLiteClusterMessenger’ as ‘this’ argument discards qualifiers [-fpermissive]
  148 |         WaitForReadyResult result = waitForReady(fdspec, command.timeout());
      |                                                                           ^
sqlitecluster/SQLiteClusterMessenger.cpp:29:44: note:   in call to ‘SQLiteClusterMessenger::WaitForReadyResult SQLiteClusterMessenger::waitForReady(pollfd&, uint64_t)’
   29 | SQLiteClusterMessenger::WaitForReadyResult SQLiteClusterMessenger::waitForReady(pollfd& fdspec, uint64_t timeoutTimestamp) {
      |                                            ^~~~~~~~~~~~~~~~~~~~~~
sqlitecluster/SQLiteClusterMessenger.cpp:198:51: error: passing ‘const SQLiteClusterMessenger’ as ‘this’ argument discards qualifiers [-fpermissive]
  198 |         if (waitForReady(fdspec, command.timeout()) != WaitForReadyResult::OK) {
      |                                                   ^
sqlitecluster/SQLiteClusterMessenger.cpp:29:44: note:   in call to ‘SQLiteClusterMessenger::WaitForReadyResult SQLiteClusterMessenger::waitForReady(pollfd&, uint64_t)’
   29 | SQLiteClusterMessenger::WaitForReadyResult SQLiteClusterMessenger::waitForReady(pollfd& fdspec, uint64_t timeoutTimestamp) {
      |                                            ^~~~~~~~~~~~~~~~~~~~~~
```

When I try `static`, it complains about accessing member functions inside of this one. Any idea how to fix or just leave it non-const/static?

Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteClusterMessenger.h:61]: It looks like you can't call `waitForReady` from a `const` function because `waitForReady` is not a `const` function. But I think `waitForReady` can be `static` and that alleviates this issue.

File: sqlitecluster/SQLiteClusterMessenger.cpp
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteClusterMessenger.cpp:90]: `data.at("name")`

**Comment by rafecolton** [comment]
@tylerkaraszewski all feedback addressed

File: libstuff/SMultiHostSocketPool.h
Comment (tylerkaraszewski) [inline libstuff/SMultiHostSocketPool.h:18]: I think this should just be `map<string, SSocketPool> _pools;` which would be simpler. I think this should work without worrying about duplicating the pool or anything by using `emplace` to insert new socket pools into `_pools`.

**Review by tylerkaraszewski — CHANGES_REQUESTED** [review]
Just one more suggested change.

**Comment by rafecolton** [comment]
@tylerkaraszewski final feedback item addressed. Marking `[HOLD]`, will merge tomorrow or Monday once approved. @bondydaa if you want to review, now's your chance 🙂

**Pull request merged**
